src/NUnitEngine/nunit-agent/Program.cs

```diff
@@ -153,16 +153,6 @@ public static int Main(string[] args)
                 log.Error("Exception in RemoteTestAgent", ex);
             }
 
-            try
-            {
-                // Unregister the channel
-                testAgencyServer.Stop();
```

Comment:
@CharliePoole
> I wasn't aware of that. We should examine what was fixed and find another way. The agent is the client in this relationship, not the server. Testagency uses Server Base so any fix there that was done for the agent should probably be removed.
> 
> Another way we made our own bug into a feature.

I don't follow. Are you saying that `TestAgency` should not be a `ServerBase`?

Any `ServerBase` that needs remoting will be susceptible to the SocketException any time the remote instance of the `ServerBase` is shutting down, so unless I 'm missing something, `ServerBase` is definitely where the fix belongs.

Maybe the confusion is that the same `ServerBase` type is instantiated twice- one on the server side, one on the client side. The fix definitely has to be there inside the implementation of `Stop` so that when `Stop` executes on the remote side, it doesn't close the socket too soon. When `Stop` is called on the test process side, none of the code in `Stop` actually runs in the test process; it just causes the remote process to execute it which is exactly what needs to happen- no more, no less.